### ISSUE TITLE
Align the use of literals, regex and glob on the configuration

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#6075]
On detekt we have a lot of configurations. Some of them are strings. But depending on the rule we interpret that string as a literal, a regex or a glob. That's confusing!

I know that align all of them is nearly impossible but it would be nice to have some guideline. And now with 2.0 I think is a really good moment.

- Literals are the simplest but doesn't give too much flexibility.
- Regex are the opposite. Complex but great flexibility.
- And glob is the middle grown.

Also we even have rules that allow literals and regex.

An option could be to define a new type of configuration.

Here an example for `ForbiddenImports`
```yaml
imports:
  - 'this string is a literal' 
importsRegex:
  - '^this string is a regex$' 
importsGlob:
  - 'this string is a glob' 
```

Other option is to have only one key and allow to define the how that string should be interpreted. 

I'm not sure which is the correct solution here but I think that we should do something to align this.

**Comment by atulgpt** [comment]
I am for `Other option is to have only one key and allow to define the how that string should be interpreted.` as for the other option where we allow 3 different options, how will we decide the priority between those 3? Or will we allow more than one to be defined in a rule?

**Comment by BraisGabin** [comment]
My idea was to allow multiple of them to be defined. We would merge them. We can map all of them to `List<Regex>`.

**Comment by marschwar** [comment]
Since `ForbiddenImports` uses the `ValuesWithReason` already those two concepts should be combined somehow. How about:

```yaml
  ForbiddenImports:
    imports:
      - value: 'TypeA.*'
        reason: reason A
        format: regex
      - value: TypeB
```

or 

```yaml
  ForbiddenImports:
    imports:
      - value: 'regex:TypeA.*'
        reason: reason A
      - value: TypeB
```

**Comment by cortinico** [comment]
> I like it. The only problem I see with all of these options (mine included) is discoberabilty. How would the users know about this? How can we make it clearly easy?

IMHO this can be added as a infobox in the website for every rule that supports it.

**Comment by marschwar** [comment]
While implementing this I ran into a conceptual issue I would like like your input on.

Assuming we add the ability to specify the type of value expression (literal, glob, regex) to the `ValuesWithReason`. Some rules (such as `ForbiddenMethodCall`) really only support the definition of literals. Should we provide a way for rule authors to declare which types the rule supports and show that in the documentation?

**Comment by BraisGabin** [comment]
I'm not sure if we should link this directly with `ValuesWithReason`. I would allow this for `ValuesWithReason` and normal `String`s.

**Comment by marschwar** [comment]
🤔 OK, that is also a thing to consider but how would we communicate that a certain config property string (or ValueWithReason) only supports literals? Should the rule raise a configuration exception?

**Comment by cortinico** [comment]
> Should the rule raise a configuration exception?

Yup that sounds like a valid approach

**Comment by BraisGabin** [comment]
Yes. I think that a configuration error could work in this case.
